test/manual/events/src/providers/video-element-provider.ts

```diff
@@ -0,0 +1,673 @@
+        return 0;
+    }
+
+    public setCurrentQuality(qualityLevel: number): void {
+    }
+
+    public getQualityLevels(): QualityLevel[] {
+        return [{
+            bitrate: 0,
+            label: '0',
+            width: this.videoElement.videoWidth,
+            height: this.videoElement.videoHeight
+        }];
+    }
+
+    public setCurrentAudioTrack(at: number): void {
+    }
+
+    public getCurrentAudioTrack(): number {
+        return 0;
+    }
+    public getAudioTracks(): { name: string, language: string }[] {
+        return [];
+    }
+
+    public getFullscreen(): boolean {
+        // Only return true if the video element itself is fullscreen (not the app/page/player)
+        return false;
+    }
+
+    public setFullscreen(isFullscreen: boolean): void {
+        // Request for the video element to go fullscreen, because the player is unable (iOS)
+    }
+
+    private getCurrentTime(): number {
+        const currentTime = this.videoElement.currentTime;
+        const seekRange = this.getSeekRange();
+        if (VideoElementProvider.isLive(this.videoElement.duration)) {
+            if (VideoElementProvider.isDvr(seekRange.end - seekRange.start, Math.max(this.config.minDvrWindow, 30))) {
+                return currentTime - seekRange.end;
+            }
+        }
+        return currentTime;
+    }
+
+    private getDuration(): number {
+        let duration = this.videoElement.duration;
+        // Don't sent time event on Android before real duration is known
+        if (isNaN(duration)) {
+            return 0;
+        }
+        const end = this.getSeekableEnd();
+        if (VideoElementProvider.isLive(this.videoElement.duration) && end) {
+            const seekableDuration = end - this.getSeekableStart();
+            if (VideoElementProvider.isDvr(seekableDuration, Math.max(this.config.minDvrWindow, 30))) {
+                // Player interprets negative duration as DVR
+                duration = -seekableDuration;
+            }
+        }
+        return duration;
+    }
+
+    private getSeekRange(): SeekRange {
+        const seekRange: SeekRange = {
+            start: 0,
+            end: 0
+        };
+        const seekable = this.videoElement.seekable;
+        if (seekable.length) {
+            seekRange.end = this.getSeekableEnd();
+            seekRange.start = this.getSeekableStart();
+        } else if (isFinite(this.videoElement.duration)) {
+            seekRange.end = this.videoElement.duration;
+        }
+        return seekRange;
+    }
+
+    private getSeekableStart(): number {
+        let start = Infinity;
+        ['buffered', 'seekable'].forEach(range => {
+            const timeRange = this.videoElement[range];
+            let index = timeRange ? timeRange.length : 0;
+            while (index--) {
+                const rangeStart = Math.min(start, timeRange.start(index));
+                if (isFinite(rangeStart)) {
+                    start = rangeStart;
```

Comment:
expect this to utilize `seekable` as priority over `buffered`?

Reply:
Neither have priority. We want the minimum of `seekable` and `buffered` start times. This corrects browser issues with seekable range either not being present or not encompassing the buffered range (which is always playable and can be seeked to).

Reply:
oh, i missed the use of `start` in the `Math.min` call on my first pass, apologies!